Dandiset permissions

AlmightyYakob opened the pull request:
Fixes #335.

Draft PR because I'm still trying to get tests working for this, but am running into some errors. Once I get those figured out I'll push those changes.

mgrauer commented:
A few issues here for consideration.

- girder already has a notion of owner, which is tied to the `AccessType.ADMIN` (I think that's the right variable name), if you use `AccessType.WRITE` for owner, that can be confusing. i'd recommend using `AccessType.ADMIN` here. But also this is specifying the user's access to the group, you haven't yet given the group access to the folder (see next point). So really what I'm saying is that the group or user should have ADMIN access on the folder.
- the dandisets are nested folders in girder. i don't see here where the group is given ownership of the dandiset folder. also, when the group is attached to the folder, it needs to have the permissions attached to the folder recursively (so that all subfolders are also owned by the group). you'll want to add a group to owning the folder at the time the dandiset is created (a different endpoint), unless you don't use groups at all.
- you don't have to use groups, could just use individual user permissions added and removed from the folder. using groups adds indirection and also requires additional machinery (code such as the functions you've added). i think groups are a positive when managing resource access by a human, i don't see them as valuable when the groups are automatically created per dandiset and have people added to them, as in this case. i'd be curious about other opinions here.

AlmightyYakob commented:
> * you don't have to use groups, could just use individual user permissions added and removed from the folder. using groups adds indirection and also requires additional machinery (code such as the functions you've added). i think groups are a positive when managing resource access by a human, i don't see them as valuable when the groups are automatically created per dandiset and have people added to them, as in this case. i'd be curious about other opinions here.

Ahh, that makes more sense, I wasn't aware you could update the access of a folder this way. I'll switch away from using groups in favor of this. This should also fix the other 2 issue you brought up, so long as I have a way to specify an admin level of access when updating the ACL of a folder.

mgrauer commented on girder-dandi-archive/girder_dandi_archive/util.py:
```diff
@@ -23,3 +25,40 @@ def validate_dandiset_identifier(dandiset_identifier):
+    return Folder().findOne({"parentId": drafts["_id"], "meta.dandiset.identifier": identifier})
+
+
+def find_dandiset(query):
```
why is this added? doesn't seem like it is used. also, what form should `query` take?

AlmightyYakob replied:
This is a general helper function which may be useful to add into several places in `rest.py`, but I haven't done that yet as I wanted consensus  from people. I encountered a weird bug from not including the `{"parentId": drafts["_id"], ...}` portion of a mongo query. This would just abstract that out, and `query` would be any valid mongo query.

mgrauer commented on girder-dandi-archive/girder_dandi_archive/rest.py:
```diff
@@ -69,21 +73,84 @@ def create_dandiset(self, params):
+            (str(user["id"]), user["level"])
+            for user in Folder().getFullAccessList(dandiset)["users"]
+        }
+        owners_to_delete = {(user["_id"], AccessType.ADMIN) for user in owners}
```
should you do some validation of passed in list of users here and in other endpoints?

AlmightyYakob replied:
I had a similar thought, but I'm not sure what I'd filter it on?

AlmightyYakob replied:
However if you mean validating that the passed in user objects themselves are valid user objects then I agree, that's something I'm going to add.

AlmightyYakob replied:
94b1e48

mgrauer commented on girder-dandi-archive/girder_dandi_archive/rest.py:
```diff
@@ -69,21 +73,84 @@ def create_dandiset(self, params):
+        Folder().requireAccess(dandiset, user=self.getCurrentUser(), level=AccessType.ADMIN)
+
+        # Make sure the list doesn't contain duplicates
+        current_owners = [
```
you aren't checking access levels in current owners, here or in other endpoints. could that have undesired effects?

it definitely means that `final_owners` could be mislabeled as it might include non-owners, which would be confusing.

AlmightyYakob replied:
Since the call to `setAccessList` will replace entirely what's currently in the ACL, we have to include all users in the ACL in `final_owners`, in order to preserve any existing users, regardless of their access level.

AlmightyYakob replied:
However I agree, the names are a bit misleading in that regard, so I can change them.

mgrauer commented on girder-dandi-archive/girder_dandi_archive/rest.py:
```diff
@@ -69,21 +73,84 @@ def create_dandiset(self, params):
+    )
+    @dandiset_identifier
+    def get_dandiset_owners(self, identifier, params):
+        return Folder().getFullAccessList(find_dandiset_by_identifier(identifier))["users"]
```
you should only return those with `ADMIN` access

AlmightyYakob replied:
Good catch, addressed in 753cce0

dchiquit commented on girder-dandi-archive/girder_dandi_archive/rest.py:
```diff
@@ -69,21 +73,84 @@ def create_dandiset(self, params):
+        new_owners = [(user["_id"], AccessType.ADMIN) for user in owners]
+        final_owners = [
+            {"id": user_id, "level": level}
+            for user_id, level in set([*current_owners, *new_owners])
```
If a user already has non-admin permissions on the folder and is added again with this method, that user will appear twice in the owners list with two different permissions levels.

Why not just use `user["_id"]` alone for `current_owners` and `new_owners` and do `final_owners = [{"id": user_id, "level": AccessType.ADMIN}]? This will upgrade any existing users to ADMIN, but that's currently what we expect 

If you're trying to preserve user permission levels, I would build a dict `{user_id: access_level}` of the current owners, add the new owners, then map that back to the required form of `[{"id":..., "level":...}]`. I think you did something similar in `remove_dandiset_owners`.

dchiquit commented on girder-dandi-archive/tests/rest/test_dandiset_ownership.py:
```diff
@@ -0,0 +1,88 @@
+    resp = server.request(path=f"/dandi/{identifier}/owners", method="GET")
+
+    assertStatusOk(resp)
+    assert len(resp.json) == 1
```
I think it's worth asserting something about the one user, even if it's just an ID.

dchiquit commented on girder-dandi-archive/tests/rest/test_dandiset_ownership.py:
```diff
@@ -0,0 +1,88 @@
+    assert len(resp.json) == 1
+
+
+def test_add_dandiset_owners(server, admin, user, admin_created_dandiset):
+    """Assert that adding a user to the list of owners produces the expected result."""
```
I don't like to assume that the response from an operation means that it changed the things it said it did. I would like to see an attempt from the user to use the folder without permission that fails with a 403, then that same attempt succeeding after being granted permission.

dchiquit commented on girder-dandi-archive/tests/rest/test_dandiset_ownership.py:
```diff
@@ -0,0 +1,88 @@
+
+
+@pytest.fixture
+def multi_owner_dandiset(server, admin, user, admin_created_dandiset):
```
It's not an issue as this currently stands, but using the `multi_owner_dandiset` fixture will corrupt the `admin_created_dandiset` fixture. I can't picture a single test that would need both, but it feels like a trap. Considering this fixture is only used in one test, it might be better to remove this fixture and just add the permissions in the test? I don't feel very strongly.

AlmightyYakob replied:
The main reason I did that is because the remove test basically contained the entirety of the add test, and I didn't like that pattern. 

I was curious about this, as I had thought that fixtures were created new and independently for each test, so I tried testing it. What I did was replace the `admin_created_dandiset` argument in `test_initial_dandiset_owners` with `multi_owner_dandiset`, and then ran the tests. The tests run in the order of:

1. `test_initial_dandiset_owners`
2. `test_add_dandiset_owners`
3. `test_remove_dandiset_owners`

I didn't replace the fixture argument of `admin_created_dandiset` in `test_add_dandiset_owners`, so if it were to corrupt the `admin_created_dandiset` fixture, in that test it should show two users as being owners on the dandiset, instead of just 1. However, when I did this, only the 1 user showed.

Furthermore if I just printed the drafts collection in each test, I get different ID each time. I know this is a lot more than you asked for, but from looking around online and testing myself, it seems like mutation of the fixtures won't actually occur, so I'd be fine leaving it as is. If there's something here I'm missing please lmk.

dchiquit replied:
Sorry, I should have been more clear. The clobbering will occur when you have a test that requests both the `multi_owner_dandiset` fixture and the `admin_created_dandiset` at the same time, e.g.
```
def broken_test(admin_created_dandiset, multi_owner_dandiset):
  assert admin_created_dandiset["_id"] != multi_owner_dandiset["_id"] # fails
```
Both fixtures refer to the same dandiset, which now has multiple owners. The way it's written now will work perfectly for any reasonable tests that I can think of. I'm not 100% happy with the way this is written, but all the alternatives I can think of have flaws of similar magnitude. Up to you.

AlmightyYakob replied:
Ahh I see what you mean. In that case I agree they should probably be independent. It will result in some duplicate code but I'm okay with that. I'll make that change.

dchiquit commented on girder-dandi-archive/tests/rest/test_dandiset_ownership.py:
```diff
@@ -0,0 +1,88 @@
+
+    users = resp.json["users"]
+    assert len(users) == 1
+    assert not [new_user for new_user in users if new_user["id"] == str(user["_id"])]
```
Same comment about asserting that the deleted user actually has the correct permissions by making some requests with the user.

AlmightyYakob commented:
This PR is definitely still a work in progress, as I'm addressing your concerns @mgrauer I'm finding some more edge cases to keep track of. I'll mark this as ready for review when I think I've addressed everything.

AlmightyYakob commented:
Had a discussion with @mgrauer, we decided that since we're going for a simplified permission model, users will either be owners of a dandiset, or have no permissions. That means that any operations done by these endpoints should be assumed to remove all existing permissions lower than `AccessType.ADMIN`.

dchiquit commented on girder-dandi-archive/tests/rest/test_dandiset_ownership.py:
```diff
@@ -0,0 +1,234 @@
+
+    users = resp.json
+    assert len(users) == 1
+    assert not [new_user for new_user in users if new_user["id"] == str(user["_id"])]
```
```suggestion
    assert resp.json[0]["id"] == str(admin["_id"])
```
I personally find list comprehensions difficult to parse. I think if we've already asserted that there is exactly one user, it is more legible to assert that that one user is the admin.

dchiquit commented on girder-dandi-archive/tests/rest/test_dandiset_ownership.py:
```diff
@@ -0,0 +1,234 @@
+    )
+
+
+def test_user_removed_from_owners(server, user, user_2, admin, admin_created_dandiset):
+    """Test that an existing user without admin permissions has none after another user is added."""
+    access_param = json.dumps(
+        {"users": [{"id": user["_id"], "level": AccessType.WRITE}]}, default=str
```
```suggestion
@pytest.mark.parametrize("accessLevel", [AccessType.READ, AccessType.WRITE])
def test_user_removed_from_owners(server, user, user_2, admin, admin_created_dandiset, accessLevel):
    """Test that an existing user without admin permissions has none after another user is added."""
    access_param = json.dumps(
        {"users": [{"id": user["_id"], "level": accessLevel}]}, default=str
```
Write one get one free
Same deal for the next test

AlmightyYakob commented on girder-dandi-archive/girder_dandi_archive/rest.py:
```diff
@@ -69,21 +75,107 @@ def create_dandiset(self, params):
+        # Make sure the list doesn't contain duplicates
+        # Only work with admin level users, removing non-admins from the ACL
+        user_id_to_level = {
+            str(user["id"]): user["level"]
+            for user in Folder().getFullAccessList(dandiset)["users"]
+            if user["level"] == AccessType.ADMIN
```
Good suggestion. As far as the difference between those two keys, `"id"` is just how the ACL of a folder returns the id of a user, instead of `"_id"` normally found elsewhere.

AlmightyYakob commented on girder-dandi-archive/girder_dandi_archive/rest.py:
```diff
@@ -69,21 +75,107 @@ def create_dandiset(self, params):
+                raise ValidationException("All owners must be valid user objects.")
+
+            # Prevents a KeyError if owner isn't an existing owner
+            user_id_to_level.pop(owner["_id"], None)
```
I originally had that, but I realized that will cause a `KeyError` if any of the users passed into the endpoint aren't already owners on the dandiset. We could handle this case and throw an error, but I'm in favor of keeping this endpoints idempotent, which this currently is.

dchiquit replied:
I had not thought of that, good point

mgrauer commented:
Can you wait to merge until I've had a chance to review?

AlmightyYakob commented:
> Can you wait to merge until I've had a chance to review?

Sure thing

mgrauer commented on girder-dandi-archive/girder_dandi_archive/util.py:
```diff
@@ -23,3 +28,46 @@ def validate_dandiset_identifier(dandiset_identifier):
+    return Folder().findOne({"parentId": drafts["_id"], "meta.dandiset.identifier": identifier})
+
+
+def dandiset_find(query, **kwargs):
```
can you remove this? it isn't called anywhere and it isn't clear to me how to call it or why it is needed from reading the code (I'm not saying it is not useful, but it is opaque to me when to call it and how, and that's from reading it in context of the PR that adds it. in the future someone coming in will not know how to use it, but will be afraid to remove it). i think it just adds cruft to the codebase and violates YAGNI.

mgrauer commented on girder-dandi-archive/girder_dandi_archive/rest.py:
```diff
@@ -69,21 +75,103 @@ def create_dandiset(self, params):
+        Folder().requireAccess(dandiset, user=self.getCurrentUser(), level=AccessType.ADMIN)
+
+        # Make sure the list doesn't contain duplicates
+        # Only work with admin level users, removing non-admins from the ACL
```
This is surprising enough that I think it deserves more explanation. Perhaps:

# Only work with admin level users, removing non-admins from the ACL/
# This is done intentionally as a simplifying assumption of providing this API.

mgrauer commented on girder-dandi-archive/girder_dandi_archive/rest.py:
```diff
@@ -69,21 +75,103 @@ def create_dandiset(self, params):
+        dandiset = find_dandiset_by_identifier(identifier)
+        Folder().requireAccess(dandiset, user=self.getCurrentUser(), level=AccessType.ADMIN)
+
+        # Only work with admin level users, removing non-admins from the ACL
```
add same comment as above, for extra explanation

mgrauer commented:
ok, a couple minor requests and then could be merged. i don't need to review further.

brianhelba commented on girder-dandi-archive/tests/rest/test_dandiset_ownership.py:
```diff
@@ -0,0 +1,231 @@
+
+
+@pytest.fixture
+def user_2(server):
```
Would it be helpful to move some of these fixtures to `conftest.py` in case they're reused in other test files, or is that not likely?

AlmightyYakob replied:
b631ba6
